python/lib/externalrequests/dmod/externalrequests/duplex/handler.py

```diff
@@ -0,0 +1,1104 @@
+            typed_handlers: Message handlers for messages that can be deserialized into a FieldedMessage
+            untyped_handlers: Message handlers for messages that can't be deserialized into Fielded messages
+            *args:
+            **kwargs:
+
+        Returns:
+            A response reporting on the success of this operation
+        """
+        # Make sure that the source is a websocket protocol and open so that messages may be correctly listened to
+        if not isinstance(source, WebSocketCommonProtocol):
+            raise ValueError(
+                f"The websocket connection to the server passed to `listen_to_messages` is not unusable; "
+                f"a `WebSocketServerProtocol` was expected, but received "
+                f"{'Nothing' if source is None else source.__class__.__name__} instead"
+            )
+        elif source.closed:
+            raise ValueError(
+                f"The websocket connection to the server passed to `listen_to_messages` is not unusable; "
+                f"it is not open."
+            )
+
+        # Make sure that the target is a websocket protocol and open so that messages may be correctly transmitted
+        if target is None:
+            logging.warning(
+                f"No target connection was passed to {common.get_current_function_name(parent_name=True)}; "
+                f"Any attempt to communicate with another service will fail. "
+            )
+        elif not isinstance(target, WebSocketCommonProtocol):
+            raise ValueError(
+                f"The websocket connection to the target passed to `listen_to_messages` is not usable; "
+                f"a `WebSocketCommonProtocol` was expected, but received "
+                f"{'Nothing' if target is None else target.__class__.__name__} instead"
+            )
+        elif target.closed:
+            raise ValueError(
+                f"The websocket connection to the target passed to `listen_to_messages` is not usable; "
+                f"it is not open."
+            )
+
+        response_data = ResponseData()
+
+        try:
+            # Listen and act upon messages that come from the source of the messages
+            async for message in source:
+                response_data += 1
+                parsed_message = self.parse_socket_input(message)
+
+                # Identify what sort of handlers to use
+                if type(parsed_message) in typed_handlers:
+                    # If a specific type of message is supposed to be used, go ahead and use those due to specificity
+                    handlers = typed_handlers[type(parsed_message)]
+                    payload = parsed_message
+                else:
+                    handlers = untyped_handlers
+                    payload = message
+
+                for handler in handlers:
+                    try:
+                        # TODO: Should these be added as tasks to run at the same time?
```

Comment:
I was kind of thinking the same thing.

Reply:
Also, if we are awaiting here, that means we cannot receive any other messages, correct? We might want to consider how to avoid blocking incoming messages while we handle a received message.

Reply:
We'll need to revisit the design. I'm going to add a todo because you have a good point, but it's easier said than done - there's already two threads associated with each duplex handler, so we don't want to be too naive with creating new groups of threads. What _might_ work would be to have an intermediary map that maps handler names to the async results. Once all have been launched, await the group. Once the group is done, assign the results to the response data.

Is it ok if we put it on the back burner for now? These external request handlers are just playing hot potato in their current state.

Reply:
Yeah, I completely agree about back burnering this for now. Whatever we decide will definitely add some complexity like you are alluding to.